common/pinot/pinotQueryValidator.go

```diff
@@ -0,0 +1,271 @@
+	// any string that surrounded by single quote as SQLVal
+	_, ok = comparisonExpr.Right.(*sqlparser.SQLVal)
+	if !ok { // this means, the value is a string, and not surrounded by single qoute, which means, val = missing
+		colVal, ok := comparisonExpr.Right.(*sqlparser.ColName)
+		if !ok {
+			return "", fmt.Errorf("error: Failed to convert val")
+		}
+		colValStr := colVal.Name.String()
+
+		// double check if val is not missing
+		if colValStr != "missing" {
+			return "", fmt.Errorf("error: failed to convert val")
+		}
+
+		var newColVal string
+		if strings.ToLower(colNameStr) == "historylength" {
+			newColVal = "0"
+		} else {
+			newColVal = "-1" // -1 is the default value for all Closed workflows related fields
+		}
+		comparisonExpr.Right = &sqlparser.ColName{
+			Metadata:  colName.Metadata,
+			Name:      sqlparser.NewColIdent(newColVal),
+			Qualifier: colName.Qualifier,
+		}
+	}
+
+	// For this branch, we still have a sqlExpr type. So need to use a buf to return the string
+	comparisonExpr.Format(buf)
+	return buf.String(), nil
+}
+
+func (qv *VisibilityQueryValidator) processCustomKey(expr sqlparser.Expr) (string, error) {
+	comparisonExpr := expr.(*sqlparser.ComparisonExpr)
+
+	colName, ok := comparisonExpr.Left.(*sqlparser.ColName)
+	if !ok {
+		return "", errors.New("invalid comparison expression, left")
+	}
+
+	colNameStr := colName.Name.String()
+
+	// check type: if is IndexedValueTypeString, change to like statement for partial match
+	valType, ok := qv.validSearchAttributes[colNameStr]
+	if !ok {
+		return "", fmt.Errorf("invalid search attribute")
+	}
+
+	// get the column value
+	colVal, ok := comparisonExpr.Right.(*sqlparser.SQLVal)
+	if !ok {
+		return "", errors.New("invalid comparison expression, right")
+	}
+	colValStr := string(colVal.Val)
+
+	// get the value type
+	indexValType := common.ConvertIndexedValueTypeToInternalType(valType, log.NewNoop())
+
+	// Case2-1: when it is string, need partial match
+	if indexValType == types.IndexedValueTypeString {
+		// change to like statement for partial match
+		comparisonExpr.Operator = sqlparser.LikeStr
+		comparisonExpr.Right = &sqlparser.SQLVal{
+			Type: sqlparser.StrVal,
+			Val:  []byte("%" + colValStr + "%"),
+		}
+		//return fmt.Sprintf("JSON_EXTRACT_SCALAR(Attr, '$.%s', 'STRING') LIKE '%%%s%%'", colNameStr, colValStr), nil
+		return fmt.Sprintf("(JSON_MATCH(Attr, '\"$.%s\" is not null') "+
+			"AND REGEXP_LIKE(JSON_EXTRACT_SCALAR(Attr, '$.%s', 'string'), '%s*'))", colNameStr, colNameStr, colValStr), nil
+	}
+	// case2-2: otherwise, exact match
+	// case2-2-1: if it is keyword, need to deal with a situation when value is an array
+	if indexValType == types.IndexedValueTypeKeyword {
+		return fmt.Sprintf("(JSON_MATCH(Attr, '\"$.%s\"=''%s''') or JSON_MATCH(Attr, '\"$.%s[*]\"=''%s'''))",
+			colNameStr, colValStr, colNameStr, colValStr), nil
+	}
+	// case2-2-2: other cases:
+	return fmt.Sprintf("JSON_MATCH(Attr, '\"$.%s\"=''%s''')", colNameStr, colValStr), nil
```

Comment:
Do we support Datetime customer key range query? From this, it doesn't see so.

Reply:
For the customized search attribtues, we don't support range queries even for integer columns. @bowenxia will test next week to see if we can use similar SQL queries to partial match query we used. [CDNC-5946](https://t3.uberinternal.com/browse/CDNC-5946)

Reply:
Update: I tested range queries and it worked. Need to update the query validator and query builder to support that. Bowen will add in a following diff.